content/blog/2019-12-20-Internal-Function-Memoization/index.md

```diff
@@ -0,0 +1,175 @@
+
+Consider the layout of recursive and iterative fibonacci programs below.
+
+<img src="fiblay.png" style="max-width: 100%" >
+
+Here, we separate the base cases from the recursive calls in order to know which set of instructions will need to be in a loop, and which set will only need to be called once. It is important to note that in the black box, you can have any set of instructions to combine the recursive calls.
+
+Consider the layout of the iterative version of the same program below.
+
+<img src="memo.png" style="max-width: 100%" >
+
+Note: We keep the function header the same in order to more easily swap out invocations to fibonacci outside the function with our generated function.
+
+### New Code Generation
+To minimize changes to the overall code structure, we delete all blocks and instructions inside the original recursive function and insert our new instructions.
+
+First we add all the base case conditionals. Next, we declare and initialize all the values we need for iteration and the incrementing iterator. After that, we create the while loop and clone in the instructions dependent on the original recursive calls (e.g., the add instruction in fib(n-1) + fib(n-2)). Lastly, we add the return statementment to which the while loop exits. 
+
+Our code for the LLVM pass is available [here](https://github.com/liuhenry4428/llvm-pass-skeleton/tree/noauto).
+
+<!-- We start of by 
+#### Fibonacci
+Consider the fibonacci algorithm. To compute number $n$, you only need the numbers (n-1) and (n-2). Any additional previous calculations can be thrown away. Is it possible to infer this information when given a program, such that we can alter a program to decrease its space consumption? If so, what are the cases to make this more generalizable, beyond optimizing a fiboncci program?
+
+Consider the call tree for recursive fibonacci.
+![](https://i.imgur.com/xaaXLDV.png)
+
+
+![](https://i.imgur.com/uCtIcSt.png)
+
+How extensible are these ideas? Consider the arbitrary one-dimensional programming problem below. Without specific knowledge of which recursive calls are required and their offsets from $n$, we must conservatively estimate a next call could use any previously computed elements.
+
+## Evaluation
+### Correctness
+In order to evaluate correctness, we look at the programs generated for the set of valid benchmarks. During construction, we separate the domain by case into separate functions, each with a different result. These _separate functions_ are the base cases, each with their own code that exeutes and returns within the case. The \{domain $\setminus$ the domain of the base cases\} is input into the loop in the body of our function. Determining program equivalence can be broken down into equivalence of these separate functions.
```

Comment:
I'm a little confused about this:

> we separate the domain by case into separate functions, each with a different result.

I don't quite see why there needs to be a separation like this, or how it's accomplished (it's not described in the program analysis section above). I thought that every base case was incorporated into one big iterative function at the end?